File: algorithms/sorting-algorithms/bubble-sort/bubble_sort.scala
val arr = Array(8,2,4,3)

def bubbleSort[T <: Ordered[T]](arr: Array[T]): Array[T] = {
    for (i <- 0 until arr.length - 1) 
    {
        var swapped = false
        for (j <- 0 until arr.length - 1 - i) 
        {
            if (arr(j) > arr(j + 1)) 
            {
                val temp = arr(j)
                arr(j) = arr(j + 1)
                arr(j + 1) = temp
                swapped = true
            }
        }
        if (!swapped) return arr
    }
    arr
}

File: algorithms/sorting-algorithms/bubble-sort/bubble_sort.hs
bubbleSort :: Ord a => [a] -> [a]
bubbleSort [] = []
bubbleSort xs = let (sorted, swapped) = bubble xs
                in if swapped then bubbleSort sorted else sorted
    where
        bubble [] = ([], False)
        bubble [x] = ([x], False)
        bubble (x:y:xs)
            | x > y = let (sorted, swapped) = bubble (x:xs)
                     in (y:sorted, True)
            | otherwise = let (sorted, swapped) = bubble (y:xs)
                         in (x:sorted, swapped)
